### ISSUE TITLE
Text selection behaviour?

### ISSUE THREAD

**Issue by sirtet** [codemirror/CodeMirror#3663]
Is CodeMirror or the Operating System responsible for the behaviour of text-selection with a doubleclick?
If CodeMirror can do something about it, it would be a nice feature to have a different behaviour for some specific cases: eg. Doubleclicking a HTML entity like _&amp;nbsp;_ selects only the nbsp part, but most users would expect the full string to be handeled as a word, i.e. to be selected. Same could apply to tags or other elements, depending on the language used.

**Comment by marijnh** [comment]
CodeMirror handles doubleclicks itself, so it would be possible to improve behavior there. There's already the possibility for modes to change the characters that count as word characters using `wordChars` helper (see for example the JavaScript mode), but there's no mechanism yet for doing this kind of contextual group matching. We could allow an additional `"findWordAt"` helper to be associated with a mode, which may either return false to pass to the default behavior, or return a range pointing at the word around the given position. Does that sound like it'd address your problem? Would you be able to submit a patch yourself? (Read the `findWordAt` method in codemirror.js for inspiration.)

**Comment by mihailik** [comment]
How about introducing `wordTokens` option in addition to simpler `wordChars`? That way mode-specific `findWordAt` is not needed.

The existing `findWordAt` would be modified to consider tokens instead of characters.

The win is that no alternative parsing logic needs to be written, it all relies on the tokenisation that is proven to work.

**Comment by marijnh** [comment]
I don't think it is usually the case that these more complicated word concepts correspond directly to tokens in the mode. It is for XML entities, but not, for example, for tags.

You don't need much parsing logic to go from a line string and a position to a start/end index in cases like this, I think.

**Comment by mihailik** [comment]
XML tags use case would work as expected with my suggestion. As in pressing `Ctr-Left` in this case:
`<in|put` ->
  `|<input` weird
  `<|input` expected

Perhaps I should try implementing it, there might be bigger obstacles on that way.

**Comment by sirtet** [comment]
@mihailik , did you ever try to do this?
I just had a bunch of html source editing to do, so i thought some more about what exactly i miss:

I found that besides wanting to select intities like `&nbsp;` i mainly wished for an easy way (via n-clicking) to select tags: I had to move around html tags with or without attributes, eg. <figure class="caption">
I think for the entity-case, a doubleclick-selection would be good, as this probably really is to be seen as one word in most cases.
But i would not say the same for Tags. maybe i just want to select the class-name, maybe the whole class-attribute, maybe the whole tag.
One option could be to increase the selected range with each doubleclick (similar as a triple click selects the whole line).
Or, the selection could depend on where i doubleclick (or, triple. leave doubleclick for normal word-selection)
So: triple on the tag name would select the tag, triple on a attribute name would select the whole attribute, etc. (eg. extend to css: triple on a selector, on a declaration)
